File: src/config/skills.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { extname, join, relative } from "node:path";
import type { DiscoveryLevel, DiscoveryResult } from "./discovery.ts";

export interface Skill {
	name: string;
	description: string;
	content: string;
	frontmatter: Record<string, string>;
	source: string;
	level: DiscoveryLevel;
}

/**
 * Parse YAML-style frontmatter from a string.
 * Expects `---` delimiters. Returns empty frontmatter if none found.
 * Uses regex-based parsing — no YAML library needed.
 */
export function parseFrontmatter(content: string): {
	frontmatter: Record<string, string>;
	body: string;
} {
	const match = content.match(/^---\s*\n([\s\S]*?\n)?---\s*\n?([\s\S]*)$/);
	if (!match) {
		return { frontmatter: {}, body: content };
	}

	const rawFrontmatter = match[1] ?? "";
	const body = (match[2] ?? "").trim();
	const frontmatter: Record<string, string> = {};

	for (const line of rawFrontmatter.split("\n")) {
		const kvMatch = line.match(/^\s*([^:]+?)\s*:\s*(.+?)\s*$/);
		if (kvMatch?.[1] && kvMatch[2]) {
			frontmatter[kvMatch[1]] = kvMatch[2];
		}
	}

	return { frontmatter, body };
}

/**
 * Parse a single skill file into a Skill object.
 *
 * @param filePath - Absolute path to the .md file
 * @param namespace - Subdirectory path relative to skills root (e.g., "foo/bar")
 * @param level - The DiscoveryLevel this skill belongs to
 */
export function parseSkillFile(filePath: string, namespace: string, level: DiscoveryLevel): Skill {
	const content = readFileSync(filePath, "utf-8");
	const { frontmatter, body } = parseFrontmatter(content);

	// Derive name from path: namespace + filename without .md, separated by colons
	const basename = filePath.replace(/\.md$/, "").split("/").pop() ?? "";
	const nameParts = namespace ? [...namespace.split("/"), basename] : [basename];
	const name = nameParts.join(":");

	const description = frontmatter.description ?? `Custom skill: ${name}`;

	return {
		name,
		description,
		content: body,
		frontmatter,
		source: level.path,
		level,
	};
}

/**
 * Recursively scan a directory for .md skill files.
 * Returns Skill objects with names derived from relative paths.
 */
function scanSkillsDir(dir: string, level: DiscoveryLevel, baseDir?: string): Skill[] {
	const base = baseDir ?? dir;
	const skills: Skill[] = [];

	try {
		const entries = readdirSync(dir);
		for (const entry of entries) {
			const fullPath = join(dir, entry);
			try {
				const stat = statSync(fullPath);
				if (stat.isDirectory()) {
					skills.push(...scanSkillsDir(fullPath, level, base));
					continue;
				}
				if (extname(entry) !== ".md") continue;

				const relDir = relative(base, dir);
				const skill = parseSkillFile(fullPath, relDir, level);
				skills.push(skill);
			} catch {
				// Skip files we can't read
			}
		}
	} catch {
		// Directory unreadable — skip
	}

	return skills;
}

/**
 * Load all skills from a DiscoveryResult.
 * Skills from earlier levels (deepest .heddle/) take priority over later levels.
 * When names collide, the first occurrence wins (deepest/highest priority).
 */
export function loadSkillsFromDiscovery(discovery: DiscoveryResult): Skill[] {
	const skillMap = new Map<string, Skill>();

	for (const level of discovery.levels) {
		let skillsDir: string;
		if (level.source === "agents") {
			// .agents/skills/ — the level path IS the skills dir
			skillsDir = level.path;
		} else {
			skillsDir = join(level.path, "skills");
		}

		const skills = scanSkillsDir(skillsDir, level);
		for (const skill of skills) {
			if (!skillMap.has(skill.name)) {
				skillMap.set(skill.name, skill);
			}
		}
	}

	return [...skillMap.values()];
}

File: src/config/skills.rs
//! Skill loading from discovery levels. Mirrors `ts-src/config/skills.ts`.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use super::discovery::{DiscoveryLevel, DiscoveryResult, DiscoverySource};

#[derive(Debug, Clone)]
pub struct Skill {
    pub name: String,
    pub description: String,
    pub content: String,
    pub frontmatter: HashMap<String, String>,
    pub source: PathBuf,
    pub level_source: DiscoverySource,
}

/// Parse YAML-style frontmatter from a string. Lightweight, key:value lines
/// only — matches the TS regex-based parser.
pub fn parse_frontmatter(content: &str) -> (HashMap<String, String>, String) {
    let mut fm = HashMap::new();
    let trimmed = content.trim_start_matches('\u{feff}');

    if !trimmed.starts_with("---") {
        return (fm, content.to_string());
    }

    // Find `\n---` after the opening
    let after_open = &trimmed[3..];
    // Match `---\n...\n---\n?`
    if let Some(close_idx) = after_open.find("\n---") {
        let raw_fm = &after_open[..close_idx];
        let body_start = close_idx + 4; // skip "\n---"
        let mut body = &after_open[body_start.min(after_open.len())..];
        // Skip optional trailing newline after the closing fence
        if let Some(s) = body.strip_prefix('\n') {
            body = s;
        }
        for line in raw_fm.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            if let Some(colon_idx) = line.find(':') {
                let key = line[..colon_idx].trim().to_string();
                let value = line[colon_idx + 1..].trim().to_string();
                if !key.is_empty() {
                    fm.insert(key, value);
                }
            }
        }
        return (fm, body.trim().to_string());
    }

    (fm, content.to_string())
}

pub fn parse_skill_file(
    file_path: &Path,
    namespace: &str,
    level: &DiscoveryLevel,
) -> Option<Skill> {
    let content = std::fs::read_to_string(file_path).ok()?;
    let (frontmatter, body) = parse_frontmatter(&content);

    let stem = file_path.file_stem().and_then(|s| s.to_str()).unwrap_or("");
    let name = if namespace.is_empty() {
        stem.to_string()
    } else {
        let mut parts: Vec<&str> = namespace.split('/').collect();
        parts.push(stem);
        parts.join(":")
    };

    let description = frontmatter
        .get("description")
        .cloned()
        .unwrap_or_else(|| format!("Custom skill: {name}"));

    Some(Skill {
        name,
        description,
        content: body,
        frontmatter,
        source: level.path.clone(),
        level_source: level.source.clone(),
    })
}

fn scan_skills_dir(dir: &Path, level: &DiscoveryLevel, base: &Path, out: &mut Vec<Skill>) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let full = entry.path();
        let ft = match entry.file_type() {
            Ok(t) => t,
            Err(_) => continue,
        };
        if ft.is_dir() {
            scan_skills_dir(&full, level, base, out);
            continue;
        }
        if full.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }
        let rel_dir = full
            .parent()
            .and_then(|p| p.strip_prefix(base).ok())
            .map(|p| p.to_string_lossy().replace('\\', "/"))
            .unwrap_or_default();
        if let Some(skill) = parse_skill_file(&full, &rel_dir, level) {
            out.push(skill);
        }
    }
}

pub fn load_skills_from_discovery(discovery: &DiscoveryResult) -> Vec<Skill> {
    let mut map: HashMap<String, Skill> = HashMap::new();
    for level in &discovery.levels {
        let skills_dir = match level.source {
            DiscoverySource::Agents => level.path.clone(),
            _ => level.path.join("skills"),
        };
        let mut levelskills = Vec::new();
        scan_skills_dir(&skills_dir, level, &skills_dir, &mut levelskills);
        for skill in levelskills {
            map.entry(skill.name.clone()).or_insert(skill);
        }
    }
    map.into_values().collect()
}
